File: implementations/python/lib/eval/king_safety.py
"""King safety evaluation - pawn shield, open files, attackers."""

from typing import Optional, Tuple
from lib.board import Board
from lib.types import Color, PieceType

PAWN_SHIELD_BONUS: int = 20
OPEN_FILE_PENALTY: int = -30
SEMI_OPEN_FILE_PENALTY: int = -15
ATTACKER_WEIGHT: int = 10


def evaluate(board: Board) -> int:
    """Evaluate king safety for both sides."""
    score = 0
    
    score += evaluate_king_safety(board, Color.WHITE)
    score -= evaluate_king_safety(board, Color.BLACK)
    
    return score


def evaluate_king_safety(board: Board, color: Color) -> int:
    """Evaluate king safety for one color."""
    king_square = find_king(board, color)
    if king_square is None:
        return 0
    
    score = 0
    
    score += evaluate_pawn_shield(board, king_square, color)
    score += evaluate_open_files(board, king_square, color)
    score -= evaluate_attackers(board, king_square, color)
    
    return score


def find_king(board: Board, color: Color) -> Optional[int]:
    """Find king square."""
    for square in range(64):
        row = square // 8
        col = square % 8
        piece = board.get_piece(row, col)
        
        if piece and piece.color == color and piece.type == PieceType.KING:
            return square
    
    return None


def evaluate_pawn_shield(board: Board, king_square: int, color: Color) -> int:
    """Evaluate pawn shield around king."""
    king_file = king_square % 8
    king_rank = king_square // 8
    shield_count = 0
    
    if color == Color.WHITE:
        shield_ranks = [king_rank + 1, king_rank + 2]
    else:
        shield_ranks = [king_rank - 1, king_rank - 2]
    
    for file in range(max(0, king_file - 1), min(8, king_file + 2)):
        for rank in shield_ranks:
            if 0 <= rank < 8:
                piece = board.get_piece(rank, file)
                if piece and piece.color == color and piece.type == PieceType.PAWN:
                    shield_count += 1
    
    return shield_count * PAWN_SHIELD_BONUS


def evaluate_open_files(board: Board, king_square: int, color: Color) -> int:
    """Evaluate open files around king."""
    king_file = king_square % 8
    penalty = 0
    
    for file in range(max(0, king_file - 1), min(8, king_file + 2)):
        own_pawns, enemy_pawns = count_pawns_on_file(board, file, color)
        
        if own_pawns == 0 and enemy_pawns == 0:
            penalty += OPEN_FILE_PENALTY
        elif own_pawns == 0:
            penalty += SEMI_OPEN_FILE_PENALTY
    
    return penalty


def count_pawns_on_file(board: Board, file: int, color: Color) -> Tuple[int, int]:
    """Count own and enemy pawns on a file."""
    own_pawns = 0
    enemy_pawns = 0
    
    for rank in range(8):
        piece = board.get_piece(rank, file)
        if piece and piece.type == PieceType.PAWN:
            if piece.color == color:
                own_pawns += 1
            else:
                enemy_pawns += 1
    
    return own_pawns, enemy_pawns


def evaluate_attackers(board: Board, king_square: int, color: Color) -> int:
    """Count enemy pieces attacking squares around king."""
    king_file = king_square % 8
    king_rank = king_square // 8
    attacker_count = 0
    
    adjacent_squares = [
        (-1, -1), (-1, 0), (-1, 1),
        (0, -1),           (0, 1),
        (1, -1),  (1, 0),  (1, 1),
    ]
    
    for dr, df in adjacent_squares:
        new_rank = king_rank + dr
        new_file = king_file + df
        
        if 0 <= new_rank < 8 and 0 <= new_file < 8:
            if is_attacked_by_enemy(board, new_rank, new_file, color):
                attacker_count += 1
    
    return attacker_count * ATTACKER_WEIGHT


def is_attacked_by_enemy(board: Board, row: int, col: int, color: Color) -> bool:
    """Check if square is attacked by enemy."""
    for attacker_square in range(64):
        attacker_row = attacker_square // 8
        attacker_col = attacker_square % 8
        piece = board.get_piece(attacker_row, attacker_col)
        
        if piece and piece.color != color:
            if can_attack(board, attacker_row, attacker_col, row, col, piece.type, piece.color):
                return True
    
    return False


def can_attack(board: Board, from_row: int, from_col: int, to_row: int, to_col: int, 
               piece_type: PieceType, color: Color) -> bool:
    """Check if piece can attack target square."""
    rank_diff = abs(to_row - from_row)
    file_diff = abs(to_col - from_col)
    
    if piece_type == PieceType.PAWN:
        forward = 1 if color == Color.WHITE else -1
        return to_row - from_row == forward and file_diff == 1
    elif piece_type == PieceType.KNIGHT:
        return (rank_diff == 2 and file_diff == 1) or (rank_diff == 1 and file_diff == 2)
    elif piece_type == PieceType.KING:
        return rank_diff <= 1 and file_diff <= 1
    
    return False

File: implementations/dart/lib/eval/king_safety.dart
import 'package:chess_engine/chess_engine.dart';

const int pawnShieldBonus = 20;
const int openFilePenalty = -30;
const int semiOpenFilePenalty = -15;
const int attackerWeight = 10;

int evaluate(Board board) {
  int score = 0;
  
  score += _evaluate_king_safety(board, PieceColor.white);
  score -= _evaluate_king_safety(board, PieceColor.black);
  
  return score;
}

int _evaluate_king_safety(Board board, PieceColor color) {
  final kingSquare = _find_king(board, color);
  if (kingSquare == null) {
    return 0;
  }
  
  int score = 0;
  
  score += _evaluate_pawn_shield(board, kingSquare, color);
  score += _evaluate_open_files(board, kingSquare, color);
  score -= _evaluate_attackers(board, kingSquare, color);
  
  return score;
}

int? _find_king(Board board, PieceColor color) {
  for (int square = 0; square < 64; square++) {
    final row = square ~/ 8;
    final col = square % 8;
    final piece = board.squares[row][col];
    
    if (piece != null && piece.color == color && piece.type == PieceType.king) {
      return square;
    }
  }
  return null;
}

int _evaluate_pawn_shield(Board board, int kingSquare, PieceColor color) {
  final kingFile = kingSquare % 8;
  final kingRank = kingSquare ~/ 8;
  int shieldCount = 0;
  
  final shieldRanks = color == PieceColor.white
      ? [kingRank + 1, kingRank + 2]
      : [(kingRank - 1).clamp(0, 7), (kingRank - 2).clamp(0, 7)];
  
  for (int file = (kingFile - 1).clamp(0, 7); file <= (kingFile + 1).clamp(0, 7); file++) {
    for (final rank in shieldRanks) {
      if (rank < 8) {
        final square = rank * 8 + file;
        final row = square ~/ 8;
        final col = square % 8;
        final piece = board.squares[row][col];
        
        if (piece != null && piece.color == color && piece.type == PieceType.pawn) {
          shieldCount++;
        }
      }
    }
  }
  
  return shieldCount * pawnShieldBonus;
}

int _evaluate_open_files(Board board, int kingSquare, PieceColor color) {
  final kingFile = kingSquare % 8;
  int penalty = 0;
  
  for (int file = (kingFile - 1).clamp(0, 7); file <= (kingFile + 1).clamp(0, 7); file++) {
    final pawns = _count_pawns_on_file(board, file, color);
    final ownPawns = pawns.$1;
    final enemyPawns = pawns.$2;
    
    if (ownPawns == 0 && enemyPawns == 0) {
      penalty += openFilePenalty;
    } else if (ownPawns == 0) {
      penalty += semiOpenFilePenalty;
    }
  }
  
  return penalty;
}

(int, int) _count_pawns_on_file(Board board, int file, PieceColor color) {
  int ownPawns = 0;
  int enemyPawns = 0;
  
  for (int rank = 0; rank < 8; rank++) {
    final square = rank * 8 + file;
    final row = square ~/ 8;
    final col = square % 8;
    final piece = board.squares[row][col];
    
    if (piece != null && piece.type == PieceType.pawn) {
      if (piece.color == color) {
        ownPawns++;
      } else {
        enemyPawns++;
      }
    }
  }
  
  return (ownPawns, enemyPawns);
}

int _evaluate_attackers(Board board, int kingSquare, PieceColor color) {
  final kingFile = kingSquare % 8;
  final kingRank = kingSquare ~/ 8;
  int attackerCount = 0;
  
  const adjacentSquares = [
    (-1, -1), (-1, 0), (-1, 1),
    (0, -1),           (0, 1),
    (1, -1),  (1, 0),  (1, 1),
  ];
  
  for (final offset in adjacentSquares) {
    final newRank = kingRank + offset.$1;
    final newFile = kingFile + offset.$2;
    
    if (newRank >= 0 && newRank < 8 && newFile >= 0 && newFile < 8) {
      final targetSquare = newRank * 8 + newFile;
      if (_is_attacked_by_enemy(board, targetSquare, color)) {
        attackerCount++;
      }
    }
  }
  
  return attackerCount * attackerWeight;
}

bool _is_attacked_by_enemy(Board board, int square, PieceColor color) {
  for (int attackerSquare = 0; attackerSquare < 64; attackerSquare++) {
    final row = attackerSquare ~/ 8;
    final col = attackerSquare % 8;
    final piece = board.squares[row][col];
    
    if (piece != null && piece.color != color) {
      if (_can_attack(board, attackerSquare, square, piece.type, piece.color)) {
        return true;
      }
    }
  }
  return false;
}

bool _can_attack(Board board, int from, int to, PieceType pieceType, PieceColor color) {
  final fromRank = from ~/ 8;
  final fromFile = from % 8;
  final toRank = to ~/ 8;
  final toFile = to % 8;
  final rankDiff = (toRank - fromRank).abs();
  final fileDiff = (toFile - fromFile).abs();
  
  switch (pieceType) {
    case PieceType.pawn:
      final forward = color == PieceColor.white ? 1 : -1;
      return toRank - fromRank == forward && fileDiff == 1;
    case PieceType.knight:
      return (rankDiff == 2 && fileDiff == 1) || (rankDiff == 1 && fileDiff == 2);
    case PieceType.king:
      return rankDiff <= 1 && fileDiff <= 1;
    default:
      return false;
  }
}
